src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-configuration-form/rbd-configuration-form.component.ts

```diff
@@ -0,0 +1,147 @@
+  }>;
+  @Output()
+  changes = new EventEmitter<any>();
+  ngDataReady = new EventEmitter<any>();
+  initialData: RbdConfigurationEntry[];
+  configurationType = RbdConfigurationType;
+  sectionVisibility: { [key: string]: boolean } = {};
+
+  constructor(
+    public formatterService: FormatterService,
+    public rbdConfigurationService: RbdConfigurationService
+  ) {}
+
+  ngOnInit() {
+    const configFormGroup = this.createConfigurationFormGroup();
+    this.form.addControl('configuration', configFormGroup);
+
+    // Listen to changes and emit the values to the parent component
+    configFormGroup.valueChanges.subscribe(() => {
+      this.changes.emit(this.getDirtyValues.bind(this));
+    });
+
+    if (this.initializeData) {
+      this.initializeData.subscribe((data) => {
+        this.initialData = data.initialData;
+        const dataType = data.sourceType;
+
+        this.rbdConfigurationService.getWritableOptionFields().forEach((option) => {
+          const optionData = data.initialData.filter((entry) => entry.name === option.name).pop();
+          if (optionData && optionData['source'] === dataType) {
+            this.form.get(`configuration.${option.name}`).setValue(optionData['value']);
+          }
+        });
+        this.ngDataReady.emit();
+      });
+    }
+
+    this.rbdConfigurationService
+      .getWritableSections()
+      .forEach((section) => (this.sectionVisibility[section.class] = false));
+  }
+
+  getDirtyValues(includeLocalValues = false, localFieldType?: RbdConfigurationSourceField) {
+    const result = {};
+
+    this.rbdConfigurationService.getWritableOptionFields().forEach((config) => {
+      const control = this.form.get('configuration').get(config.name);
+      const dirty = control.dirty;
+
+      if (this.initialData && this.initialData[config.name] === control.value) {
+        return; // Skip controls with initial data loaded
+      }
+
+      if (dirty || (includeLocalValues && control['source'] === localFieldType)) {
+        if (control.value === null) {
+          result[config.name] = control.value;
+        } else if (config.type === RbdConfigurationType.bps) {
+          result[config.name] = this.formatterService.toBytes(control.value);
+        } else if (config.type === RbdConfigurationType.milliseconds) {
+          result[config.name] = this.formatterService.toMilliseconds(control.value);
+        } else if (config.type === RbdConfigurationType.iops) {
+          result[config.name] = this.formatterService.toIops(control.value);
+        } else {
+          result[config.name] = control.value;
+        }
+      }
+    });
+
+    return result;
+  }
+
+  /**
+   * Dynamically create form controls.
+   */
+  private createConfigurationFormGroup() {
+    const configFormGroup = new CdFormGroup({});
+
+    this.rbdConfigurationService.getWritableOptionFields().forEach((c) => {
+      let control: FormControl;
+      if (
+        c.type === RbdConfigurationType.milliseconds ||
```

Comment:
That would disable the check for valid types and defeat the purpose of the code.

Reply:
I just found out you can do the following with enums:

`if (c.type in RbdConfigurationType)`

This would not change the behavior and would be much shorter :)